be/src/exprs/ngram.cpp

```diff
@@ -0,0 +1,215 @@
+        }
+    }
+
+    ColumnPtr static haystack_const_and_needle_const(const ColumnPtr& haystack_ptr, const Slice& needle) {
+        std::unique_ptr<NgramHash[]> hash(new NgramHash[map_size]);
+        memset(hash.get(), 0, sizeof(NgramHash) * map_size);
+
+        const Slice& haystack = ColumnHelper::get_const_value<TYPE_VARCHAR>(haystack_ptr);
+        // if haystack is too large, we can say they are not similar at all
+        if (haystack.get_size() >= max_string_size) {
+            return ColumnHelper::create_const_column<TYPE_FLOAT>(0, haystack_ptr->size());
+        }
+
+        Slice cur_haystack(haystack.get_data(), haystack.get_size());
+
+        std::string buf;
+        if constexpr (case_insensitive) {
+            tolower(haystack, buf);
+            cur_haystack = Slice(buf.c_str(), buf.size());
+        }
+        // needle_gram_count may be zero because needle is empty or N is too large for needle
+        size_t needle_gram_count = calculateMapWithNeedle(hash.get(), needle);
+        size_t needle_not_overlap_with_haystack =
+                calculateDistanceWithHaystack<false>(hash.get(), cur_haystack, nullptr, needle_gram_count);
+        float result = 1.0f - (needle_not_overlap_with_haystack)*1.0f / std::max(needle_gram_count, (size_t)1);
+        DCHECK(needle_not_overlap_with_haystack <= needle_gram_count);
+        return ColumnHelper::create_const_column<TYPE_FLOAT>(result, haystack_ptr->size());
+    }
+
+private:
+    void inline static tolower(const Slice& str, std::string& buf) {
+        buf = str.to_string();
+        std::transform(buf.begin(), buf.end(), buf.begin(), [](unsigned char c) { return std::tolower(c); });
+    }
+};
+
+template <size_t N, bool case_insensitive, bool use_utf_8, class Gram>
+StatusOr<ColumnPtr> static ngram_search_impl(const Columns& columns) {
+    RETURN_IF_COLUMNS_ONLY_NULL(columns);
+    const ColumnPtr& haystack_ptr = columns[0];
+
+    const ColumnPtr& needle_ptr = columns[1];
+
+    if (!needle_ptr->is_constant()) {
+        return Status::NotSupported("ngram function's second parameter must be const");
+    }
+
+    const Slice& needle = ColumnHelper::get_const_value<TYPE_VARCHAR>(needle_ptr);
+    if (needle.get_size() >= max_string_size) {
+        return Status::NotSupported("ngram function's second parameter is larger than 2^15");
+    }
+
+    // needle is too small so we can not get even single Ngram, so they are not similar at all
+    if (needle.get_size() < N) {
+        return ColumnHelper::create_const_column<TYPE_FLOAT>(0, haystack_ptr->size());
+    }
+
+    if (haystack_ptr->is_constant()) {
+        return NgramFunctionImpl<N, case_insensitive, use_utf_8, Gram>::haystack_const_and_needle_const(haystack_ptr,
+                                                                                                        needle);
+    } else {
+        return NgramFunctionImpl<N, case_insensitive, use_utf_8, Gram>::haystack_vector_and_needle_const(haystack_ptr,
```

Comment:
(S(needle) ∪ S(haystack) - S(needle)∩S(haystack))  /  (|needle| + | haystack|)

for S(needle), S(needle), |needle| we only need to compute once